tonic-types/src/richer_error/mod.rs

```diff
@@ -0,0 +1,500 @@
+use prost::{DecodeError, Message};
+use prost_types::Any;
+use tonic::{codegen::Bytes, metadata::MetadataMap, Code};
+
+mod error_details;
+mod std_messages;
+
+use super::pb;
+
+pub use error_details::{vec::ErrorDetail, ErrorDetails};
+pub use std_messages::{BadRequest, FieldViolation, RetryInfo};
+
+trait IntoAny {
+    fn into_any(self) -> Any;
+}
+
+trait FromAny {
+    fn from_any(any: Any) -> Result<Self, DecodeError>
+    where
+        Self: Sized;
+}
+
+fn gen_details_bytes(code: Code, message: &String, details: Vec<Any>) -> Bytes {
+    let status = pb::Status {
+        code: code as i32,
+        message: message.clone(),
+        details,
+    };
+
+    Bytes::from(status.encode_to_vec())
+}
+
+/// Used to implement associated functions and methods on `tonic::Status`, that
+/// allow the addition and extraction of standard error details.
+pub trait StatusExt {
```

Comment:
We probably want to seal this trait similar to https://docs.rs/tower-http/latest/src/tower_http/builder.rs.html#45 that way outside users can't implement it and we can add more required fn as we please without making a breaking change.

Reply:
You are right. I did not know this pattern, quite useful. Thanks!